python/cuml/neighbors/nearest_neighbors_mg.pyx

```diff
@@ -0,0 +1,350 @@
+
+    cdef cppclass RankSizePair:
+        int rank
+        size_t size
+
+    cdef cppclass PartDescriptor:
+        PartDescriptor(size_t M,
+                       size_t N,
+                       vector[RankSizePair*] &partsToRanks,
+                       int myrank)
+
+cdef extern from "cumlprims/opg/selection/knn.hpp" namespace \
+        "MLCommon::Selection::opg":
+
+    cdef void brute_force_knn(
+        cumlHandle &handle,
+        vector[int64Data_t*] &out_I,
+        vector[floatData_t*] &out_D,
+        vector[floatData_t*] &idx_data,
+        PartDescriptor &idx_desc,
+        vector[floatData_t*] &query_data,
+        PartDescriptor &query_desc,
+        bool rowMajorIndex,
+        bool rowMajorQuery,
+        int k,
+        size_t batch_size,
+        bool verbose
+    ) except +
+
+
+class NearestNeighborsMG(NearestNeighbors):
+    """
+    Multi-node multi-GPU Nearest Neighbors kneighbors query.
+
+    NOTE: This implementation of NearestNeighbors is meant to be
+    used with an initialized cumlCommunicator instance inside an
+    existing distributed system. Refer to the Dask NearestNeighbors
+     implementation in `cuml.dask.neighbors.nearest_neighbors`.
+    """
+    def __init__(self, batch_size=1<<21, **kwargs):
+        super(NearestNeighborsMG, self).__init__(**kwargs)
+        self.batch_size = batch_size
+
+    def _build_dataFloat(self, arr_interfaces):
+        """
+        Instantiate a container object for a float data pointer
+        and size.
+        :param arr_interfaces:
+        :return:
+        """
+        cdef vector[floatData_t*] *dataF = new vector[floatData_t*]()
+
+        cdef uintptr_t input_ptr
+        for x_i in range(len(arr_interfaces)):
+            x = arr_interfaces[x_i]
+            input_ptr = x["data"]
+            data = <floatData_t*> malloc(sizeof(floatData_t))
+            data.ptr = <float *> input_ptr
+            data.totalSize = < size_t > (x["shape"][0] *
+                                         x["shape"][1] *
+                                         sizeof(float))
+
+            dataF.push_back(data)
+
+        return <size_t>dataF
+
+    def _free_mem(self, index_vec, index_desc,
+                  query_vec, query_desc,
+                  out_i_vec, out_d_vec,
+                  local_index_parts,
+                  local_query_parts):
+
+        cdef vector[floatData_t *] *index_vec_c \
+            = <vector[floatData_t *]*><size_t>index_vec
+        cdef PartDescriptor *index_desc_c \
+            = <PartDescriptor*><size_t>index_desc
+        for elm in range(index_vec_c.size()):
+            free(index_vec_c.at(elm))
+        free(index_vec_c)
+        free(index_desc_c)
+
+        cdef vector[floatData_t *] *query_vec_c \
+            = <vector[floatData_t *]*><size_t>query_vec
+        cdef PartDescriptor *query_desc_c \
+            = <PartDescriptor*><size_t>query_desc
+        for elm in range(query_vec_c.size()):
+            free(query_vec_c.at(elm))
```

Comment:
could _freeFloatD be used here? I confess it's a little hard to keep track of all the array frees and a higher-level wrapper like freeFloatD would make it easier to read.

Reply:
Good catch. I think I might be able to use that here.